source/language/openpulse.rst

```diff
@@ -0,0 +1,823 @@
+
+.. code-block:: javascript
+
+  ...
+
+   // Example 1: Shift phase of the "drive" frame by pi/4, to realize a virtual rz gate with angle -pi/4
+   cal {
+     driveframe.phase += pi/4;
+   }
+
+   // The following is an example only. Frames as arrays has not been agreed on.
+   // This conceptually must be compile-time arrays and treat qubits as indices
+   // which also has not been well-defined. We are exploring other solutions to
+   // the problem of mapping qubits to pulse-level resources.
+
+   // Example 2: Define a calibration for the rz gate on all 8 physical qubits
+   cal {
+     array[frame, 8] rz_frames;
+     frame[0] = newframe(...);
+     // and so on
+   }
+
+   defcal rz(angle[20] theta) $q {
+     rz_frames[q].phase -= theta;
+   }
+
+Manipulating frames based on the state of other frames is also permitted:
+
+.. code-block:: javascript
+
+   // Swap phases between two frames
+   const temp = frame1.phase;
+   frame1.phase = frame2.phase;
+   frame2.phase = temp;
+
+Waveforms
+---------
+
+Waveforms are of type ``waveform`` and can either be:
+
+- An array of complex samples which define the points for the waveform envelope
+- An abstract mathematical function representing a waveform. This will later be
+  materialized into a list of complex samples, either by the compiler or the hardware
+  using the parameters provided to the ``extern`` defined pulse template.
+
+A value of type ``waveform`` may be defined either by explicitly constructing the complex samples
+or by calling one of the waveform template functions provided by the target device.
+Note that each of these extern functions takes a type ``duration`` as an argument,
+since waveforms must have a definite duration.
+Using the hardware dependent ``dt`` unit is recommended for this duration,
+since otherwise the compiler may need to down-sample a higher precision
+waveform to physically realize it.
+
+Like other extern functions, ``extern waveform`` functions will be compiled.
+But for static waveforms, the optimizing compiler should decide to execute this
+at compile time and load the waveform into memory once.
+For dynamic waveforms, the compiler just compiles and links this, to be executed at runtime.
+We provide the ``waveform`` type in addition to the complex list of samples to
+provide more context to compilers and hardware. For example, some hardware pulse
+generators may have optimized implementations of common pulse shapes like gaussians.
+Providing structured gaussian parameters instead of the materialized list of complex
+samples provides optimization opportunities that wouldn't be available otherwise.
+
+.. code-block:: javascript
+
+   // arbitrary complex samples
+   waveform arb_waveform = [1+0*j, 0+1*j, 1/sqrt(2)+1/sqrt(2)*j];
+
+   // amp is waveform amplitude at center
+   // d is the overall duration of the waveform
+   // sigma is the standard deviation of waveform
+   extern gaussian(complex[size] amp, duration d, duration sigma) -> waveform;
```

Comment:
Here, and basically in all the example below, these `complex` types need to be `complex[float[size]]`.